elf_load.c

```diff
@@ -311,13 +330,233 @@ static int elf_load_imgcpy(struct elf_prog *elf_prog, Elf *elf,
 }
 
 #if CONFIG_LIBVFSCORE
+#if CONFIG_LIBPOSIX_MMAP
+/* If vastart + phdr.p_filesz (vastart) < vastart + phdr.p_memsz (vaend),
+ * 0 out that remainder, either through memset or through anonymous mappings
+ */
+static int elf_load_mmap_filesz_memsz_diff(struct elf_prog *elf_prog,
+					   GElf_Phdr *phdr __maybe_unused,
+					   uintptr_t vastart, uintptr_t vaend)
+{
+	UK_ASSERT(elf_prog && phdr && vastart && vaend);
+
+	uk_pr_debug("%s: Zeroing 0x%"PRIx64" - 0x%"PRIx64"\n",
+		    elf_prog->name,
+		    (uint64_t)(vastart),
+		    (uint64_t)(vaend));
+
+	/* From the last byte contained in filesz to the last
+	 * byte contained in memsz, we can either have:
+	 * 1. a page alignment adjustment, where the end
+	 * address is just PAGE_ALIGN_UP(paddr + filesz), which
+	 * means that our initial call to mmap already read in
+	 * a page for us whose remaining bytes we memset without
+	 * generating a page fault, because
+	 * vaend - vastart < PAGE_SIZE
+	 * ...
+	 */
+	memset((void *)vastart, 0, PAGE_ALIGN_UP(vastart) - vastart);
+
+	if (vaend == PAGE_ALIGN_UP(vastart))
+		return 0;
+
+	/*
+	 * ...
+	 * 2. vaend - vastart >= PAGE_SIZE, which can happen if
+	 * this segment contains a NOBITS section, such as .bss.
+	 * If that is the case, then simply anonymously map this
+	 * remaining area so that we don't waste time memsetting
+	 * it (.bss is quite large).
+	 */
+	vastart = PAGE_ALIGN_UP(vastart);
+	vastart = (uintptr_t)mmap((void *)vastart, vaend - vastart,
+				  PROT_EXEC | PROT_READ | PROT_WRITE,
+				  MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS,
+				  -1, 0);
+	if (unlikely(vastart == (uintptr_t)MAP_FAILED)) {
+		uk_pr_err("Failed to mmap the NOBITS part of phdr at "
+			  "offset %lu\n", phdr->p_offset);
+		return (int)vastart;
+	}
+
+	return 0;
+}
+
+/* Use this to mmap first PT_LOAD */
+static int do_elf_load_fdphdr_0(struct elf_prog *elf_prog,
+				GElf_Phdr *phdr, int fd)
+{
+	uintptr_t vastart_old, vastart_aligned;
+	uintptr_t vaend_old, vaend_new;
+	uintptr_t vastart, vaend;
+	__sz mmap_len;
+	int rc;
+
+	/* First start/vabase can't be !0 before loading first PT_LOAD */
+	UK_ASSERT(!elf_prog->start && !elf_prog->vabase && phdr);
+	/* PT_LOAD p_vaddr/p_paddr must be multiple of page size */
+	UK_ASSERT(PAGE_ALIGNED(phdr->p_vaddr));
```

Comment:
If we align the first segment with `pt_load_max_align`, let's assert `phdr->p_vaddr % pt_load_max_align == 0`. Otherwise there will be empty pages in front of the first segment. This should not break anything if the `v_addr` of the first segment is 0 (in common cases).